modcc/solvers.cpp

```diff
@@ -528,6 +529,258 @@ void LinearSolverVisitor::finalize() {
+            auto temp_dvar_term = make_unique_local_assign(e->scope(), dvar_ident.get(), "t_");
+            auto temp_dvar = temp_dvar_term.id->is_identifier()->spelling();
+
+            // Save the variable names
+            dvar_init_.push_back(init_dvar);
+            dvar_temp_.push_back(temp_dvar);
+
+            // Add local declarations and assignment statements
+            statements_.push_back(std::move(init_dvar_term.local_decl));
+            statements_.push_back(std::move(init_dvar_term.assignment));
+
+            statements_.push_back(std::move(temp_dvar_term.local_decl));
+            statements_.push_back(std::move(temp_dvar_term.assignment));
+        }
+    }
+    scale_factor_.resize(dvars_.size());
+
+    BlockRewriterBase::visit(e);
+}
+
+void SparseNonlinearSolverVisitor::visit(CompartmentExpression *e) {
+    auto loc = e->location();
+
+    for (auto& s: e->is_compartment()->state_vars()) {
+        auto it = std::find(dvars_.begin(), dvars_.end(), s->is_identifier()->spelling());
+        if (it == dvars_.end()) {
+            error({"COMPARTMENT variable is not used", loc});
+            return;
+        }
+        auto idx = it - dvars_.begin();
+        scale_factor_[idx] = e->scale_factor()->clone();
+    }
+}
+
+void SparseNonlinearSolverVisitor::visit(AssignmentExpression *e) {
+    if (A_.empty()) {
+        unsigned n = dvars_.size();
+        A_ = symge::sym_matrix(n, n);
+    }
+
+    auto loc = e->location();
+    scope_ptr scope = e->scope();
+
+    auto lhs = e->lhs();
+    auto rhs = e->rhs();
+    auto deriv = lhs->is_derivative();
+
+    if (!deriv) {
+        statements_.push_back(e->clone());
+
+        auto id = lhs->is_identifier();
+        if (id) {
+            auto expand = substitute(rhs, local_expr_);
+            if (involves_identifier(expand, dvars_)) {
+                local_expr_[id->spelling()] = std::move(expand);
+            }
+        }
+        return;
+    }
+
+    auto s = deriv->name();
+    auto expanded_rhs = substitute(rhs, local_expr_);
+
+    // Populate sparse symbolic matrix for GE.
+    if (s!=dvars_[deq_index_]) {
+        error({"ICE: inconsistent ordering of derivative assignments", loc});
+        return;
+    }
+
+    auto dt_expr = make_expression<IdentifierExpression>(loc, "dt");
+    auto one_expr = make_expression<NumberExpression>(loc, 1.0);
+
+    // Form and save F(x) = x(t+1) - x(t) - dt(G(x(t+1))
+    // x(t+1) are stored in dvar_temp_ and are updated at every iteration of Newton's method
+    // x(t)   are stored in dvar_init_ and are constant across iterations of Newton's method
+    // G(x)   is the rhs of the derivative assignment expression
+
+    expression_ptr F_x;
+    F_x = make_expression<MulBinaryExpression>(loc, expanded_rhs->clone(), dt_expr->clone());
+    if (scale_factor_[deq_index_]) {
+        F_x = make_expression<DivBinaryExpression>(loc, std::move(F_x), scale_factor_[deq_index_]->clone());
```

Comment:
There's the possibility of a small improvement in performance if the reciprocals of the scale factors are assigned to associated variables, which are multiplied here, rather than taking the division each time.

Reply:
I agree, I had initially assumed these would typically be identifier expressions, but that's not necessarily true.